Explicitly list all types in factories for friendly error messages

timofeymukha opened the pull request:
This is a suggestion to enhance all our json-based factories with a KNOWN_TYPES list with the names of all the types the factory can create. We can then output this list when an unknown type is provided by the user in the case file.

This has proven to be a nice way of exploring what is available in the code in e.g. OpenFOAM. One intentionally puts a jibberish string in the "type" just to see the options. The minus is a bit of extra burden on the developers, but I think it is tolerable.

So far "don't merge" as I will extend it to all factories upon approval.

Question: is the string concatenation routine ok? Will it just reallocate the result to the necessary size as we are concating in new items to it?

njansson commented on src/simulation_components/simulation_component_factory.f90:
```diff
@@ -42,12 +42,21 @@ module simulation_component_fctry
   public :: simulation_component_factory
 
+  ! List of all possible types created by the factory routine
+  character(len=20) :: KNOWN_TYPES(5) = [character(len=20) :: &
+     "vorticity", &
+     "lambda2", &
+     "probes", &
+     "les_model", &
+     "field_writer"]
+
```
Is the extra `[character...` necessary?

In other cases it works without, for example:  https://gitlab.com/ExtremeFLOW/neko_opr/-/blob/master/driver.f90?ref_type=heads#L19

```suggestion
  character(len=20) :: KNOWN_TYPES(5) = [&
     "vorticity", &
     "lambda2", &
     "probes", &
     "les_model", &
     "field_writer"]

```

timofeymukha replied:
Interesting, in my case it would crash saying the strings are of different lengths. Can it be that the example you link to works because you pad all the stings to the same length with spaces?

timofeymukha commented:
Besides for the `character` thing, are we in agreement for extending this to other types?

timfelle reviewed (commented):
> Besides for the `character` thing, are we in agreement for extending this to other types? 

It quite a good idea to have this. One think I thougfht of was, what is the maximal string length and could that be an issue?

timofeymukha commented:
@njansson @timfelle @MartinKarp The PR is now open for review. In addtion to adding the KNOWN_TYPES, in indroduced the following name convention:
- The name of the variable for the object being allocated is `object`
- The name of the variable for the name of the type, (extracted from the json or passed to the factory) is `type_name`

njansson commented:
There's several changes in the BSD license text, probably from some refactoring, which needs to be reverted

njansson reviewed (changes requested):
BSD license text shall not be altered

timofeymukha commented:
> BSD license text shall not be altered

I knew using %s/ will bite me in the butt :-)

timfelle commented on src/krylov/krylov_fctry.f90:
```diff
@@ -46,213 +46,236 @@ module krylov_fctry
        type is(sx_cg_t)
-          call kp%free()
+          call obj%free()
        type is(cg_device_t)
-          call kp%free()
+          call obj%free()
        type is(pipecg_t)
-          call kp%free()
+          call obj%free()
        type is(sx_pipecg_t)
-          call kp%free()
+          call obj%free()
        type is (pipecg_device_t)
-          call kp%free()
+          call obj%free()
        type is (fusedcg_device_t)
-          call kp%free()
+          call obj%free()
        type is(cacg_t)
-          call kp%free()
+          call obj%free()
        type is(gmres_t)
-          call kp%free()
+          call obj%free()
        type is(sx_gmres_t)
-          call kp%free()
+          call obj%free()
        type is(gmres_device_t)
-          call kp%free()
+          call obj%free()
        type is(bicgstab_t)
-          call kp%free()
+          call obj%free()
        end select
     end if
 
```
Why are we not just calling `obejct%free()`

timofeymukha replied:
@njansson ?

timfelle replied:
We do it in stuff like the simcomps

njansson replied:
Should work, but we had issues with compilers in the past

timfelle replied:
Interesting, which compilers did you try out, and did you find out what went wrong? From what i understand it should be functional in terms of the Fortran language. So is it just missing features in some of the compilers?

timfelle replied:
> I think more like bugs than missing features :-)

Oh the joy 😹

timofeymukha replied:
I mean the death of ifx here is a good example :-). Myabe something on our side, but then why don't all the other compilers complain. And anyway, ifx just segfaulted instead of complaining about the code ^^

timfelle commented on src/krylov/krylov_fctry.f90:
```diff
@@ -46,213 +46,236 @@ module krylov_fctry
-       select type(kp => ksp)
+       select type(obj => object)
        type is(cg_t)
-          call kp%init(n, max_iter)
+          call obj%init(n, max_iter)
        type is(sx_cg_t)
-          call kp%init(n, max_iter)
+          call obj%init(n, max_iter)
        type is(cg_device_t)
-          call kp%init(n, max_iter)
+          call obj%init(n, max_iter)
        type is(pipecg_t)
-          call kp%init(n, max_iter)
+          call obj%init(n, max_iter)
        type is(sx_pipecg_t)
-          call kp%init(n, max_iter)
+          call obj%init(n, max_iter)
        type is (pipecg_device_t)
-          call kp%init(n, max_iter)
+          call obj%init(n, max_iter)
        type is (fusedcg_device_t)
-          call kp%init(n, max_iter)
+          call obj%init(n, max_iter)
        type is(cacg_t)
-          call kp%init(n, max_iter)
+          call obj%init(n, max_iter)
        type is(gmres_t)
-          call kp%init(n, max_iter)
+          call obj%init(n, max_iter)
        type is(sx_gmres_t)
-          call kp%init(n, max_iter)
+          call obj%init(n, max_iter)
        type is(gmres_device_t)
-          call kp%init(n, max_iter)
+          call obj%init(n, max_iter)
        type is(bicgstab_t)
-          call kp%init(n, max_iter)
+          call obj%init(n, max_iter)
        end select
     end if
```
Is there any good reason we do this massive init case stuff. Can't we just do
init, without selecting the type? The type is detected through the allocation
stuff above.

njansson replied:
Same here, should work, but we had issues with compilers in the past.

njansson replied:
Fully agree, but given the experience, not too keen on introducing these PITA compiler issues at this time

timofeymukha replied:
Let us at this point add a comment for why it is there, and that we should at some point see if we can get away with removing this stuff.

timfelle replied:
Sure, I fully support attempting that kind of change in a separate PR with rigorous testing if we already know there might be issues. Its not relevant for the scope of this PR anyway. 

Maybe we should add an issue to track this?

timofeymukha replied:
Done

timfelle commented on src/krylov/precon_fctry.f90:
```diff
@@ -41,37 +41,52 @@ module precon_fctry
+    character(len=*), intent(in) :: type_name
+    character(len=:), allocatable :: type_string
+
+    type_string =  concat_string_array(KNOWN_TYPES, NEW_LINE('A') // "-  ", &
+                                       .true.)
 
```
Maybe move this to the else statement, since we dont use it before then

timofeymukha replied:
True

timfelle commented on src/les/les_model_fctry.f90:
```diff
@@ -14,7 +14,7 @@
 !     with the distribution.
 !
-!   * Neither the name of the authors nor the names of its
+!   * Neither the type_name of the authors nor the names of its
```
```suggestion
!   * Neither the name of the authors nor the names of its
```

timofeymukha commented:
@njansson NB: ifx just got fixed by commenting out the `private` statements I added to the precon and ksp factory modules. I think we should try to clean up the `use` statements with respect to these types, there seems to be a lot of implicit propagation going on via just `using` the factory module.

njansson commented:
> @njansson NB: ifx just got fixed by commenting out the `private` statements I added to the precon and ksp factory modules. I think we should try to clean up the `use` statements with respect to these types, there seems to be a lot of implicit propagation going on via just `using` the factory module.

Great, but what a mess (sigh). I through that we had cleaned these in the past.

timofeymukha commented:
> > @njansson NB: ifx just got fixed by commenting out the `private` statements I added to the precon and ksp factory modules. I think we should try to clean up the `use` statements with respect to these types, there seems to be a lot of implicit propagation going on via just `using` the factory module.
> 
> Great, but what a mess (sigh). I through that we had cleaned these in the past.

We did to some extent, but not thoroughly enough perhaps :-). Although, for individual ksp types the `use` seems to be pretty damn clean, almost everything behind `only.`

timfelle commented on src/.depends:
I updated it
